libvisual/libvisual/private/lv_video_blit.cpp

```diff
@@ -61,30 +61,37 @@ namespace LV {
       }
   }
 
-  void VideoBlit::blit_overlay_alphasrc (Video* dest, Video* src)
+  void VideoBlit::blit_overlay_alphasrc (Video* dst, Video* src)
   {
-      auto destbuf = static_cast<uint8_t*> (dest->get_pixels ());
-      auto srcbuf  = static_cast<uint8_t const*> (src->get_pixels ());
-
       if (visual_cpu_has_mmx ()) {
-          blit_overlay_alphasrc_mmx (dest, src);
+          blit_overlay_alphasrc_mmx (dst, src);
           return;
       }
 
+      auto dst_pixel_row_ptr = static_cast<uint8_t*> (dst->get_pixels ());
+      auto src_pixel_row_ptr = static_cast<uint8_t const*> (src->get_pixels ());
+
       for (int y = 0; y < src->m_impl->height; y++) {
+          auto dst_pixel = dst_pixel_row_ptr;
+          auto src_pixel = src_pixel_row_ptr;
+
           for (int x = 0; x < src->m_impl->width; x++) {
-              uint8_t alpha = srcbuf[3];
+              uint8_t const src_alpha = src_pixel[3];
+
+              uint16_t const c0 = static_cast<uint16_t> (src_pixel[0]) * src_alpha + static_cast<uint16_t> (dst_pixel[0]) * (255 - src_alpha);
+              uint16_t const c1 = static_cast<uint16_t> (src_pixel[1]) * src_alpha + static_cast<uint16_t> (dst_pixel[1]) * (255 - src_alpha);
+              uint16_t const c2 = static_cast<uint16_t> (src_pixel[2]) * src_alpha + static_cast<uint16_t> (dst_pixel[2]) * (255 - src_alpha);
 
-              destbuf[0] = (alpha * (srcbuf[0] - destbuf[0]) >> 8) + destbuf[0];
-              destbuf[1] = (alpha * (srcbuf[1] - destbuf[1]) >> 8) + destbuf[1];
-              destbuf[2] = (alpha * (srcbuf[2] - destbuf[2]) >> 8) + destbuf[2];
+              dst_pixel[0] = c0 >> 8;
+              dst_pixel[1] = c1 >> 8;
+              dst_pixel[2] = c2 >> 8;
```

Comment:
Hi @kaixiong ,

the TL;DR version would be that I think that:
- The `>> 8` in `dst_pixel[0] = c0 >> 8;` and siblings is too much
- The cast to 16 bits needs to happen elsewhere, i.e. not `cast(a * b + cast(c * d))` but `cast(a * b) + cast(c * d)`.

Let me go into more detail on former:

My impression is that this tries to implement approach "A over B" at https://en.wikipedia.org/wiki/Alpha_compositing#Description , formalized there as:

```
A over B

a_0 := a_a + a_b * (1.0 - a_a)
C_0 := (C_a * a_a + C_b * a_b * (1.0 - a_a)) / a_0
```

Whereas for us:

```
A := src
B := dst
```

With `a_a` and `a_b` ranging from 0 to 255 rather than 0.0 to 1.0 I end up with these formulas for us, not the output range annotation.

```
a_0 := a_a/255 + a_b/255 * (1.0 - a_a/255))  # range 0..1
    == a_a/255 + a_b/255 * (255/255 - a_a/255))
    == (a_a + a_b * (255/255 - a_a/255)))/ 255
    == (a_a + a_b * (255 - a_a) / 255) / 255  # range 0..1
    => (a_a + a_b * (255 - a_a) / 255)        # range 0..255

C_0 := (C_a * a_a/255 + C_b * a_b/255 * (1.0 - a_a/255)) / (a_0/255)  # range 0..1
    == (C_a * a_a/255 + C_b * a_b/255 * (1.0 - a_a/255)) / a_0 * 255
    == (C_a * a_a/255 + C_b * a_b/255 * (255/255 - a_a/255)) / a_0 * 255
    == (C_a * a_a/255 + C_b * a_b/255 * ((255 - a_a) / 255)) / a_0 * 255  # range 0..1
    => (C_a * a_a     + C_b * a_b     * ((255 - a_a) / 255)) / a_0 * 255  # range 0..255
```    

Now the current code in the pull request seems to assume that the `dst` image has an alpha channel with full 255 opacity `a_b` for all pixels.  I'd like to understand why and would like to suggest addition of a comment to the code, but I'll take it for granted below for a moment.

So Inserting `a_b := 255` I get …

```
a_0 := (a_a + a_b * (255 - a_a) / 255)
    == (a_a + 255 * (255 - a_a) / 255)
    == (a_a + (255 - a_a))
    == 255

C_0 := (C_a * a_a + C_b * a_b * ((255 - a_a) / 255)) / a_0 * 255
    == (C_a * a_a + C_b * 255 * ((255 - a_a) / 255)) / 255 * 255
    == (C_a * a_a + C_b *       ((255 - a_a)      ))
    == (src_pixel[0] * src_alpha + dst_pixel[0] * ((255 - src_alpha)))
```

Due to `a_0 == 255` then `dst_pixel[3]` does not need to be written, confirmed.
However, `C_0` is already in range 0..255 so the additional division in `dst_pixel[0] = c0 >> 8;` seems to be to much and should yield in an "almost black" picture, if I am not mistaken.

I'm happy to learn what I missed or to take this to a call, e.g. if my side here was hard to understand.

Best, Sebastian

Reply:
@hartwork, just to be clear this is only meant to be a fix of the original code. The operation here doesn't actually implement 'over' (or 'atop'), it's a simple linear interpolation between the source and target colours, using the source alpha as parameter. The target alpha is unchanged.

At some point, I want to have alpha compositing with the complete set of operators but that'll have to come later. There are additional API design considerations in there due to the necessity of premultiplied alpha, which the Wikipedia article mentions.

Regarding the actual calculation itself, the right-shift by 8 bits is necessary. Alpha is a percentage i.e. value in [0.0, 1.0] that's been mapped to [0, 255]. So what we're calculating here is really:

```
c = c1 * alpha/255 + c2 * (1 - alpha/255)
```
Multiplying throughout by 255 to work with only integers, we have:
```
c * 255 = c1 * alpha + c2 * (255 - alpha)
```

To get back `c`, we divide by 255 in the final step. Here` >> 8` was originally chosen for performance reasons (as the test mentions), so I kept it as it is.

Since `c1`, `c2` and `alpha` each have a maximum value of 255 and their products are involved, we need to work in at least `uint16_t`.

Reply:
@kaixiong thanks for elaborating, let me digest that more.